pex/interpreter_constraints.py

```diff
@@ -0,0 +1,90 @@
+# Copyright 2014 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+# A library of functions for filtering Python interpreters based on compatibility constraints
+
+from .interpreter import PythonIdentity
+
+
+def _matches(interpreter, filters, meet_all_constraints=False):
+  if meet_all_constraints:
+    return all(interpreter.identity.matches(filt) for filt in filters)
+  else:
+    return any(interpreter.identity.matches(filt) for filt in filters)
+
+
+def _matching(interpreters, filters, meet_all_constraints=False):
+  for interpreter in interpreters:
+    if _matches(interpreter, filters, meet_all_constraints):
+      yield interpreter
+
+
+def check_requirements_are_well_formed(constraints):
+  # Check that the compatibility requirements are well-formed.
+  for req in constraints:
+    try:
+      PythonIdentity.parse_requirement(req)
+    except ValueError as e:
+      from .common import die
+      die("Compatibility requirements are not formatted properly: %s" % str(e))
+
+
+def matched_interpreters(interpreters, filters, meet_all_constraints=False):
+  """Given some filters, yield any interpreter that matches at least one of them, or all of them
+     if meet_all_constraints is set to True.
+
+  :param interpreters: a list of PythonInterpreter objects for filtering
+  :param filters: A sequence of strings that constrain the interpreter compatibility for this
+    pex, using the Requirement-style format, e.g. ``'CPython>=3', or just ['>=2.7','<3']``
+    for requirements agnostic to interpreter class.
+  :param meet_all_constraints: whether to match against all filters.
+    Defaults to matching interpreters that match at least one filter.
+  """
+  for match in _matching(interpreters, filters, meet_all_constraints):
+    yield match
+
+
+def parse_interpreter_constraints(constraints_string):
```

Comment:
I added this `parse_interpreter_constraints` method + a test when I originally implemented the interpreter constraints PEX cli option as a single comma-separated string. Upon further iteration, I changed this option to be a list that accepts multiple re-uses of the same --interpreter-constraint=X flag to construct a list of constraints (similar to declaring --python=Y multiple times) and subsequently made this method obsolete. I kept it in the PR in case we find later on that it would be good to define interpreter constraints as a single string, but AFAIK that will not be the case and I will remove this.